torch/csrc/jit/tensorexpr/analysis.h

```diff
@@ -248,6 +248,221 @@ class ModifiesVarChecker : public IRVisitor {
+// before A's ancestor or A in the block. For example, "1:1:0" points to stmt
+// "d[i] += e[i, j];" in the following code.
+//
+// c = a/b;
+// for i
+//  d[i] = 0;
+//  for j
+//    d[i] += e[i, j];
+class StmtTrace {
+ public:
+  void append(int32_t id) {
+    trace_.emplace_back(id);
+  }
+  void pop() {
+    trace_.pop_back();
+  }
+
+  const std::vector<int32_t>& getTrace() {
+    return trace_;
+  }
+
+  bool operator==(StmtTrace& trace) {
+    auto compare = trace.getTrace();
+    if (trace_.size() != compare.size()) {
+      return false;
+    }
+    int size = compare.size();
+    for (int i = 0; i < size; i++) {
+      if (trace_.at(i) != compare.at(i)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool operator<(StmtTrace& trace) {
+    auto compare = trace.getTrace();
+    int size = trace_.size() < compare.size() ? trace_.size() : compare.size();
+    for (int i = 0; i < size; i++) {
+      if (trace_.at(i) > compare.at(i)) {
+        return false;
+      }
+    }
+    return !(*this == trace);
+  }
+
+  bool operator>(StmtTrace& trace) {
+    auto compare = trace.getTrace();
+    int size = trace_.size() < compare.size() ? trace_.size() : compare.size();
+    for (int i = 0; i < size; i++) {
+      if (trace_.at(i) < compare.at(i)) {
+        return false;
+      }
+    }
+    return !(*this == trace);
+  }
+
+  std::string getTraceString() {
+    std::string cstr = "";
+    for (int i = 0; i < trace_.size(); i++) {
+      cstr += std::to_string(trace_.at(i));
+      cstr += ":";
+    }
+    return cstr;
+  }
+
+ private:
+  std::vector<int32_t> trace_;
+};
+
+// Traverse the root stmt to identify the position of each stmt in the AST,
+// i.e., a StmtTrace
+class StmtTracer : public IRVisitor {
+ public:
+  StmtTrace getStmtTrace() {
+    return trace_;
+  }
+
+ private:
+  void visit(BlockPtr v) {
+    if (flatten_) {
+      for (StmtPtr s : *v) {
+        s->accept(this);
+      }
+      return;
+    }
+
+    int count = 0;
+    for (StmtPtr s : *v) {
+      trace_.append(count);
+      s->accept(this);
+      trace_.pop();
+      count++;
+    }
+  }
+
+  void visit(CondPtr v) {
+    flatten_ = true;
+    ExprPtr condition = v->condition();
+    StmtPtr true_stmt = v->true_stmt();
+    StmtPtr false_stmt = v->false_stmt();
+    condition->accept(this);
+    if (true_stmt) {
+      true_stmt->accept(this);
+    }
+    if (false_stmt) {
+      false_stmt->accept(this);
+    }
+    flatten_ = false;
+  }
+
+  StmtTrace trace_;
+  bool flatten_ = false;
+};
+
+enum AccMode { READ, WRITE, REDUCE };
+
+// Traverses the IR to identify all reads/writes to a buf, and their positions
+// in the AST which is represented as a StmtTrace
+using BufAccessInfo = std::tuple<StmtPtr, AccMode, StmtTrace>;
+class BufAccesses : public StmtTracer {
+ public:
+  BufAccesses(BufPtr b) : buf_(b) {}
+
+  std::vector<std::tuple<StmtPtr, AccMode, StmtTrace>> accesses() {
```

Comment:
Nit: probably should be `std::vector<BufAccessInfo>`

Reply:
> Why don't we know that S1 executes before S2? Would it be better if we simply specify which branch of If the stmt belong to?

We know S1 executes before S2 but we don't know S3 executes before S4 because we'll flatten the if-stmt to one node in the AST. In this example, S1 - [0], S2 - [1], S3/S4/S5 - [2].
The reason we flatten if-stmt is that we can not differentiate if branches using a vector of integers: in current definition, an integer in the vector represents the stmt counter in a `Block`, and all stmts in a block will be executed from top to the bottom. That's why we can use the vector of integers to identify the execution order of buffer access stmts.
In the example, if S3 - [2, 0, 0], S5 - [2, 1, 0], then we will assume S5 executes after S3. One of the corresponding stmts is,

```
for {
  S1
  S2
  if {
    S3
    S4
  } 
  if { 
    S5
  }
}
```

We add extend `StmtTrace` to express cond stmts if there are use cases in the future.